lib/DocumentManagerContext.php

```diff
@@ -0,0 +1,199 @@
+<?php
+
+/*
+ * This file is part of Sulu.
+ *
+ * (c) MASSIVE ART WebServices GmbH
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Component\DocumentManager;
+
+use PHPCR\SessionInterface;
+use ProxyManager\Factory\LazyLoadingGhostFactory;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
+
+/**
+ * The document manager context is a container for all the dependencies of a
+ * given document manager.
+ */
+class DocumentManagerContext
+{
+    /**
+     * @var EventDispatcherInterface
+     */
+    private $eventDispatcher;
+
+    /**
+     * @var DocumentInspectorFactory
+     */
+    private $inspectorFactory;
+
+    /**
+     * @var NodeManager
+     */
+    private $nodeManager;
+
+    /**
+     * @var DocumentRegistry
+     */
+    private $registry;
+
+    /**
+     * @var ProxyFactory
+     */
+    private $proxyFactory;
+
+    /**
+     * @var MetadataFactoryInterface
+     */
+    private $metadataFactory;
+
+    /**
+     * @var SessionInterface
+     */
+    private $session;
+
+    /**
+     * @var DocumentManagerInterface
+     */
+    private $manager;
+
+    /**
+     * @var string
+     */
+    private $name;
+
+    public function __construct(
+        $name,
+        SessionInterface $session,
+        EventDispatcherInterface $eventDispatcher,
+        MetadataFactoryInterface $metadataFactory,
+        LazyLoadingGhostFactory $lazyProxyFactory,
+        DocumentInspectorFactoryInterface $inspectorFactory,
+        $defaultLocale
+    ) {
+        $this->name = $name;
+
+        // the event dispatcher should be unique to this document manager context
+        // instance.
+        $this->eventDispatcher = $eventDispatcher;
+
+        // the inspector factory provides a way for users to instantiate
+        // their own document inspectors.
+        $this->inspectorFactory = $inspectorFactory;
+
+        // the metadata factory is currently intended to be the same for all
+        // document mnaager contexts instances.
+        $this->metadataFactory = $metadataFactory;
+
+        // the PHPCR session SHOULD be unique to all (end user) document
+        // manager contexts.
+        $this->session = $session;
+
+        // instantiate other objects scoped to this document manager.
+        $this->nodeManager = new NodeManager($session);
+        $this->proxyFactory = new ProxyFactory($this, $lazyProxyFactory, $metadataFactory);
+        $this->registry = new DocumentRegistry($defaultLocale);
+        $this->manager = new DocumentManager($this);
```

Comment:
- Event dispatcher is a service which needs subscribers to be added to it.
- Event dispatcher could be different (DebugEventDispatcher f.e.).
- Event dispatcher has no constructor dependencies.

We could make it optional. But I don't see any advantage at this stage.

On Thu, Apr 28, 2016 at 12:32:35AM -0700, Daniel Rotter wrote:
>    In [1]lib/DocumentManagerContext.php:
> 
>  > +        // their own document inspectors.
>  > +        $this->inspectorFactory = $inspectorFactory;
>  > +
>  > +        // the metadata factory is currently intended to be the same for all
>  > +        // document mnaager contexts instances.
>  > +        $this->metadataFactory = $metadataFactory;
>  > +
>  > +        // the PHPCR session SHOULD be unique to all (end user) document
>  > +        // manager contexts.
>  > +        $this->session = $session;
>  > +
>  > +        // instantiate other objects scoped to this document manager.
>  > +        $this->nodeManager = new NodeManager($session);
>  > +        $this->proxyFactory = new ProxyFactory($this, $lazyProxyFactory, $metadataFactory);
>  > +        $this->registry = new DocumentRegistry($defaultLocale);
>  > +        $this->manager = new DocumentManager($this);
> 
>    I see, but why don't we instantiate the EventDispatcher also in here then?
> 
>    —
>    You are receiving this because you authored the thread.
>    Reply to this email directly or [2]view it on GitHub
> 
>    Reverse link: [3]unknown
> 
> References
> 
>    Visible links
>    1. https://github.com/sulu/sulu-document-manager/pull/66#discussion_r61383902
>    2. https://github.com/sulu/sulu-document-manager/pull/66/files/05556aabdb564b020fa388baf338b54d39353978#r61383902
>    3. https://github.com/sulu/sulu-document-manager/pull/66/files/05556aabdb564b020fa388baf338b54d39353978#r61383902